lib/frames.js

```diff
@@ -126,6 +126,38 @@ function defineFrame(type, definition) {
   return Frame;
 }
 
+// restricted type helpers
+function role(value) {
```

Comment:
Man, it really _feels_ like there should be a way to easily factor this into common code. I just can't see what it is (I keep seeing how I'd write it in C#).